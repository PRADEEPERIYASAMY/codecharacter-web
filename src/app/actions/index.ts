export * from 'app/actions/dashboard';
