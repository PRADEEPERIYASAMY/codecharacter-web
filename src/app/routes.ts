export enum Routes {
  ROOT = '/',
  LOGIN = '/login',
  REGISTER = '/register',
  LEADERBOARD = '/leaderboard',
  USER_PROFILE_MODEL = '/profile',
<<<<<<< HEAD
  GITHUB_OAUTH = '/login/github',
  GOOGLE_OAUTH = '/login/google',
  USER_ACTIVATION = '/user-activate',
=======
  PROFILE_USER_STATS = '/profile/:username',
>>>>>>> Add route for /profile/:user and Add the required container
}
