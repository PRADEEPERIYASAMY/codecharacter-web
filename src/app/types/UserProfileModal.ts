import * as UserInterfaces from 'app/types/User';

export interface StateProps {
  isUserProfileModalOpen: boolean;
  userDetails: UserInterfaces.UserStoreState;
}

export interface ListDisabled {
  isUserNameDisabled: boolean;
  isFullNameDisabled: boolean;
  isEmailDisabled: boolean;
  isFlagSelectDisabled: boolean;
  isPasswordDisabled: boolean;
}
export interface State {
  username: string;
  oldPassword: string;
  password: string;
  repeatPassword: string;
  email: string;
  country: string;
  fullName: string;
  listDisabled: ListDisabled;
}

export interface DispatchProps {
  checkUsernameExists: (username: string) => void;
  editUserPassword: (editUserPasswordDetails: UserInterfaces.EditUserPassword) => void;
  editUserProfile: (editUserDetails: UserInterfaces.EditUserDetails) => void;
  getUserDetails: () => void;
  toggleUserProfileModal: (isUserProfileModalOpen: boolean) => void;
}

export type Props = StateProps & DispatchProps;
