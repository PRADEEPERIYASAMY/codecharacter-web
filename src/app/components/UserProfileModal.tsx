import {
  faEnvelope,
  faFlag,
  faLock,
  faPen,
  faTimes,
  faUser,
  faUserTie,
} from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import * as styles from 'app/styles/UserProfileModal.module.css';
import * as UserProfileInterfaces from 'app/types/UserProfileModal';
import classnames from 'classnames';
import * as React from 'react';
import { Button, Grid, Row } from 'react-bootstrap';

// tslint:disable-next-line:import-name
import ReactFlagsSelect from 'react-flags-select';
import 'react-flags-select/css/react-flags-select.css';
export class UserProfileModal extends React.Component<
  UserProfileInterfaces.Props,
  UserProfileInterfaces.State
> {
  private editProfileRef = React.createRef<HTMLFormElement>();
  private editPasswordRef = React.createRef<HTMLFormElement>();
  private reactFlag = React.createRef<ReactFlagsSelect>();

  constructor(props: UserProfileInterfaces.Props) {
    super(props);
    const { userDetails } = this.props;
    this.state = {
      country: userDetails.country,
      email: userDetails.email,
      fullName: userDetails.fullName,
      listDisabled: {
        isEmailDisabled: true,
        isFlagSelectDisabled: true,
        isFullNameDisabled: true,
        isPasswordDisabled: true,
        isUserNameDisabled: true,
      },
      oldPassword: '',
      password: '',
      repeatPassword: '',
      username: userDetails.username,
    };
    this.props.getUserDetails();
  }

  public componentWillReceiveProps(nextProps: UserProfileInterfaces.Props) {
    const { country, username, fullName } = this.state;
    const { userDetails } = nextProps;
    this.setState({
      country: userDetails.country !== country ? userDetails.country : country,
      fullName: userDetails.fullName !== fullName ? userDetails.fullName : fullName,
      username: userDetails.username !== country ? username : userDetails.username,
    });
    if (this.reactFlag.current) {
      this.reactFlag.current.updateSelected(
        userDetails.country !== country ? userDetails.country : country,
      );
    }
  }

  public render() {
    const {
      email,
      oldPassword,
      password,
      fullName,
      repeatPassword,
      username,
      country,
      listDisabled,
    } = this.state;
    const {
      checkUsernameExists,
      isUserProfileModalOpen,
      toggleUserProfileModal,
      userDetails,
    } = this.props;
    return (
      <div
        className={classnames(styles.UserProfileModal, {
          [`${styles.modalOpen}`]: isUserProfileModalOpen,
        })}
      >
        <Grid
          fluid={true}
          className={classnames(styles.modal, {
            [`${styles.slideEnter}`]: isUserProfileModalOpen,
          })}
        >
          <Row>
            <div className="col" />
            <div className="col-sm-auto p-0 m-0">
              <Button
                className={classnames(styles.customBtn)}
                onClick={() => {
                  toggleUserProfileModal(false);
                }}
              >
                <FontAwesomeIcon icon={faTimes} color={'black'} />
              </Button>
            </div>
          </Row>
          <Row>
            <div className="col text-center mb-3">
              <h3 className="text-dark m-0">Edit Profile</h3>
            </div>
          </Row>
          <Row>
            <div className="col-6">
              <Row className="mb-3">
                <div
                  className={classnames('col-sm-12', styles.form)}
                  style={{
                    borderRight: '1px solid rgb(235, 235, 235)',
                    color: 'red',
                  }}
                >
                  <div className={classnames('text-dark', styles.formHeading)}>
                    {' '}
                    Basic Information{' '}
                  </div>
                  <form
                    className={'editForm'}
                    noValidate
                    ref={this.editProfileRef}
                    onSubmit={this.handleEditProfile}
                  >
                    <div className="form-row">
                      <div className="col mb-3">
                        <div className="input-group">
                          <div className="input-group-prepend">
                            <span className="input-group-text" id="inputGroupPrepend">
                              <FontAwesomeIcon icon={faUser} />
                            </span>
                          </div>
                          <input
                            type="text"
                            className="form-control"
                            id="editValidationUsername"
                            placeholder="Username"
                            aria-describedby="inputGroupPrepend"
                            maxLength={50}
                            minLength={5}
                            value={username}
                            onChange={(e) => {
                              checkUsernameExists(e.target.value);
                              this.setState({
                                username: e.target.value,
                              });
                            }}
                            required
                            disabled={listDisabled.isUserNameDisabled}
                          />
                          <div className="input-group-append">
                            <span
                              className={classnames('input-group-text', styles.editPen, {
                                [`${styles.editPenActive}`]: !listDisabled.isUserNameDisabled,
                              })}
                              onClick={() => {
                                this.setState({
                                  listDisabled: {
                                    ...listDisabled,
                                    isUserNameDisabled: !listDisabled.isUserNameDisabled,
                                  },
                                });
                              }}
                            >
                              <FontAwesomeIcon icon={faPen} />
                            </span>
                          </div>
                          <div className="invalid-feedback">
                            Username must have minimum 5 characters.
                          </div>
                        </div>
                      </div>
                    </div>
                    <div className="form-row">
                      <div className="col mb-3">
                        <div className="input-group">
                          <div className="input-group-prepend">
                            <span className="input-group-text" id="inputGroupPrepend">
                              <FontAwesomeIcon icon={faUserTie} />
                            </span>
                          </div>
                          <input
                            type="text"
                            className="form-control"
                            id="editValidationFullname"
                            placeholder="Name"
                            aria-describedby="inputGroupPrepend"
                            maxLength={50}
                            minLength={5}
                            value={fullName}
                            onChange={(e) =>
                              this.setState({
                                fullName: e.target.value,
                              })
                            }
                            required
                            disabled={listDisabled.isFullNameDisabled}
                          />
                          <div className="input-group-append">
                            <span
                              className={classnames('input-group-text', styles.editPen, {
                                [`${styles.editPenActive}`]: !listDisabled.isFullNameDisabled,
                              })}
                              onClick={() => {
                                this.setState({
                                  listDisabled: {
                                    ...listDisabled,
                                    isFullNameDisabled: !listDisabled.isFullNameDisabled,
                                  },
                                });
                              }}
                            >
                              <FontAwesomeIcon icon={faPen} />
                            </span>
                          </div>
                          <div className="invalid-feedback">
                            Name must have minimum 5 characters.
                          </div>
                        </div>
                      </div>
                    </div>
                    <div className="form-row">
                      <div className="col mb-3">
                        <div className="input-group">
                          <div className="input-group-prepend">
                            <span className="input-group-text" id="inputGroupPrepend">
                              <FontAwesomeIcon icon={faEnvelope} />
                            </span>
                          </div>
                          <input
                            type="email"
                            className="form-control"
                            id="editValidationEmail"
                            placeholder="Email"
                            aria-describedby="inputGroupPrepend"
                            value={email}
                            onChange={(e) =>
                              this.setState({
                                email: e.target.value,
                              })
                            }
                            required
                            disabled={listDisabled.isEmailDisabled}
                          />
                          <div className="input-group-append">
                            <span
                              className={classnames('input-group-text', styles.editPen, {
                                [`${styles.editPenActive}`]: !listDisabled.isEmailDisabled,
                              })}
                              onClick={() => {
                                this.setState({
                                  listDisabled: {
                                    ...listDisabled,
                                    isEmailDisabled: !listDisabled.isEmailDisabled,
                                  },
                                });
                              }}
                            >
                              <FontAwesomeIcon icon={faPen} />
                            </span>
                          </div>
                          <div className="invalid-feedback">Please enter a valid Email ID.</div>
                        </div>
                      </div>
                    </div>
                    <div className="form-row">
                      <div className="col sm={12} mb-3">
                        <div className="input-group-prepend">
                          <div className="input-group-prepend">
                            <span className="input-group-text" id="inputGroupPrepend">
                              <FontAwesomeIcon icon={faFlag} />
                            </span>
                          </div>
                          <ReactFlagsSelect
                            searchable={true}
                            placeholder="Search for a country"
                            className="customFlag"
                            defaultCountry={country}
                            onSelect={this.onSelectFlag}
                            ref={this.reactFlag}
                            disabled={listDisabled.isFlagSelectDisabled}
                          />
                          <div className="input-group-append">
                            <span
                              className={classnames('input-group-text', styles.editPen, {
                                [`${styles.editPenActive}`]: !listDisabled.isFlagSelectDisabled,
                              })}
                              onClick={() => {
                                this.setState({
                                  listDisabled: {
                                    ...listDisabled,
                                    isFlagSelectDisabled: !listDisabled.isFlagSelectDisabled,
                                  },
                                });
                              }}
                            >
                              <FontAwesomeIcon icon={faPen} />
                            </span>
                          </div>
                          <div className="invalid-feedback">Please select a country.</div>
                        </div>
                      </div>
                    </div>
                    <div className="form-row">
                      <div className="input-group" />
                      <div className="col text-center mt -0 mb-2 errorMessage">
                        {userDetails.errorMessage}
                      </div>
                    </div>
                    <div className="form-row">
                      <div className="col text-center mb-2">
                        <button
                          className="btn btn-info"
                          type="submit"
                          style={{
                            width: '100%',
                          }}
                        >
                          Edit Profile
                        </button>
                      </div>
                    </div>
                  </form>
                </div>
              </Row>
            </div>
            <div className="col-6">
              <Row className={classnames('mb-3')}>
                <div className={classnames('col-sm-12', styles.form)}>
                  <div
                    className={classnames('text-dark', styles.formHeading)}
                    style={{
                      display: 'flex',
                    }}
                  >
                    {' '}
                    Credentials
                    <Button
                      className={classnames(styles.editPen, 'ml-auto p-0', {
                        [`${styles.editPenActive}`]: !listDisabled.isPasswordDisabled,
                      })}
                      onClick={() => {
                        this.setState({
                          listDisabled: {
                            ...listDisabled,
                            isPasswordDisabled: !listDisabled.isPasswordDisabled,
                          },
                        });
                      }}
                    >
                      <FontAwesomeIcon icon={faPen} />
                    </Button>
                  </div>
                  <form
                    className={'editpasswordForm'}
                    noValidate
                    ref={this.editPasswordRef}
                    onSubmit={this.handleEditPassword}
                  >
                    <div className="form-row">
                      <div className="col mb-3">
                        <div className="input-group">
                          <div className="input-group-prepend">
                            <span className="input-group-text" id="inputGroupPrepend">
                              <FontAwesomeIcon icon={faLock} />
                            </span>
                          </div>
                          <input
                            type="password"
                            className="form-control"
                            id="editValidationOldPassword"
                            placeholder="Old Password"
                            aria-describedby="inputGroupPrepend"
                            minLength={5}
                            value={oldPassword}
                            onChange={(e) => {
                              this.setState({
                                oldPassword: e.target.value,
                              });
                            }}
                            required
                            disabled={listDisabled.isPasswordDisabled}
                          />
                          <div className="invalid-feedback">
                            Password should have minimum 5 characters.
                          </div>
                        </div>
                      </div>
                    </div>
                    <div className="form-row">
                      <div className="col mb-3">
                        <div className="input-group">
                          <div className="input-group-prepend">
                            <span className="input-group-text" id="inputGroupPrepend">
                              <FontAwesomeIcon icon={faLock} />
                            </span>
                          </div>
                          <input
                            type="password"
                            className="form-control"
                            id="editValidationPassword"
                            placeholder="New Password"
                            aria-describedby="inputGroupPrepend"
                            minLength={5}
                            value={password}
                            onChange={(e) => {
                              this.setState({
                                password: e.target.value,
                              });
                            }}
                            required
                            disabled={listDisabled.isPasswordDisabled}
                          />
                          {listDisabled.isPasswordDisabled ? (
                            <div className="invalid-feedback">Passwords should match.</div>
                          ) : null}
                        </div>
                      </div>
                    </div>
                    <div className="form-row">
                      <div className="col mb-5">
                        <div className="input-group">
                          <div className="input-group-prepend">
                            <span className="input-group-text" id="inputGroupPrepend">
                              <FontAwesomeIcon icon={faLock} />
                            </span>
                          </div>
                          <input
                            type="password"
                            className="form-control"
                            id="editValidationRepeatPassword"
                            placeholder="Confirm Password"
                            aria-describedby="inputGroupPrepend"
                            minLength={5}
                            value={repeatPassword}
                            onChange={(e) => {
                              this.setState({
                                repeatPassword: e.target.value,
                              });
                            }}
                            required
                            disabled={listDisabled.isPasswordDisabled}
                          />
                          {listDisabled.isPasswordDisabled ? (
                            <div className="invalid-feedback">Passwords should match.</div>
                          ) : null}
                        </div>
                      </div>
                    </div>
                    <div className="col-12 text-center">
                      <button
                        className="btn btn-info"
                        type="submit"
                        style={{
                          marginTop: '36px',
                          width: '100%',
                        }}
                      >
                        Edit Credentials
                      </button>
                    </div>
                  </form>
                </div>
              </Row>
            </div>
          </Row>
        </Grid>
      </div>
    );
  }

  private handleEditProfile = (event: React.FormEvent<HTMLFormElement>) => {
    const { editUserProfile } = this.props;
    const { country, email, fullName, username, listDisabled } = this.state;
    const form = this.editProfileRef.current;
    event.preventDefault();

    if (form) {
      if (form.checkValidity()) {
        editUserProfile({
          ...(!listDisabled.isFlagSelectDisabled && { country }),
          ...(!listDisabled.isEmailDisabled && { email }),
          ...(!listDisabled.isFullNameDisabled && { fullName }),
          ...(!listDisabled.isUserNameDisabled && { username }),
        });

        this.setState({
          listDisabled: {
            ...listDisabled,
            isEmailDisabled: true,
            isFlagSelectDisabled: true,
            isFullNameDisabled: true,
            isUserNameDisabled: true,
          },
        });
      }
      form.classList.add('was-validated');
    }
  };

  private handleEditPassword = (event: React.FormEvent<HTMLFormElement>) => {
    const { editUserPassword } = this.props;
    const { oldPassword, repeatPassword, password, listDisabled } = this.state;
    const form = this.editPasswordRef.current;
    event.preventDefault();

    if (form) {
      if (form.checkValidity()) {
        if (!listDisabled.isPasswordDisabled && password === repeatPassword) {
          editUserPassword({
            ...(!listDisabled.isPasswordDisabled && {
              oldPassword,
              password,
            }),
          });
        }
        this.setState({
          listDisabled: {
            ...listDisabled,
            isPasswordDisabled: true,
          },
        });
      }
      form.classList.add('was-validated');
    }
  };

  private onSelectFlag = (countryCode: string) => {
    this.setState({
      country: countryCode,
    });
  };
}
