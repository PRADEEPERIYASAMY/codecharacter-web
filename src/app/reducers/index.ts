import { codeReducer } from 'app/reducers/code/Code';
import { editorReducer } from 'app/reducers/code/Editor';
import { dashboardReducer } from 'app/reducers/Dashboard';
import { gameLogReducer } from 'app/reducers/GameLog';
import { leaderboardReducer } from 'app/reducers/Leaderboard';
import { notificationReducer } from 'app/reducers/Notification';
import { userReducer } from 'app/reducers/User';
import * as CodeInterfaces from 'app/types/code/Code';
import * as EditorInterfaces from 'app/types/code/Editor';
import * as DashboardInterfaces from 'app/types/Dashboard';
import * as GameLogInterfaces from 'app/types/GameLog';
import * as LeaderboardInterfaces from 'app/types/Leaderboard';
import * as UserInterfaces from 'app/types/User';
import { routerReducer, RouterState } from 'react-router-redux';
import { combineReducers } from 'redux';

export const rootReducer = combineReducers({
  code: codeReducer,
  dashboard: dashboardReducer,
  editor: editorReducer,
  gameLog: gameLogReducer,
  leaderboard: leaderboardReducer,
  notification: notificationReducer,
  router: routerReducer,
  user: userReducer,
});

export interface RootState {
  code: CodeInterfaces.CodeStoreState;
  dashboard: DashboardInterfaces.DashboardStoreState;
  editor: EditorInterfaces.EditorStoreState;
  leaderboard: LeaderboardInterfaces.LeaderboardStoreState;
  router: RouterState;
  gameLog: GameLogInterfaces.GameLogStoreState;
  user: UserInterfaces.UserStoreState;
}
