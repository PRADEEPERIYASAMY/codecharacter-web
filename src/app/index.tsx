import { Dashboard } from 'app/components/Dashboard';
import * as React from 'react';
import { hot } from 'react-hot-loader';
import { Route, Switch } from 'react-router';

/* tslint:disable-next-line:variable-name */
export const App = hot(module)(() => (
  <Switch>
    <Route path="/" component={Dashboard} />
  </Switch>
));
