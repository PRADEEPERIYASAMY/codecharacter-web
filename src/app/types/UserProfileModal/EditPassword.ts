import * as UserInterfaces from 'app/types/User';
import { InputName, InputState, ListDisabled } from 'app/types/UserProfileModal';
import * as React from 'react';
export interface ElementOwnProps {
  handleEditPassword: (event: React.FormEvent<HTMLFormElement>) => void;
  onInputChange: (inputName: InputName, value: string) => void;
  inputEnabler: (inputState: InputState, value: boolean) => void;
  editPasswordRef: React.Ref<HTMLFormElement>;
  listDisabled: ListDisabled;
  userDetails: UserInterfaces.UserStoreState;
  oldPassword: string;
  password: string;
  repeatPassword: string;
}

export type Props = ElementOwnProps;
