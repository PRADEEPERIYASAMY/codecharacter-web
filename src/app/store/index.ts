import { logger } from 'app/middleware';
import { rootReducer } from 'app/reducers';
import { applyMiddleware, createStore } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';

export function configureStore(initialState?: object) {
  let middleware = applyMiddleware(logger);

  if (process.env.NODE_ENV !== 'production') {
    middleware = composeWithDevTools(middleware);
  }

  const store = createStore(rootReducer, initialState, middleware);

  if (module.hot) {
    module.hot.accept('app/reducers', () => {
      const nextReducer = require('app/reducers');
      store.replaceReducer(nextReducer);
    });
  }

  return store;
}
