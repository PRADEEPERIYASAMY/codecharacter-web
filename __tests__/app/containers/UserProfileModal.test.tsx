import UserProfileModal from 'app/containers/UserProfileModal';
import { configureStore } from 'app/store';
import { shallow } from 'enzyme';
import * as React from 'react';

describe('UserProfileModal Container', () => {
  const store = configureStore();
  const wrapper = shallow(<UserProfileModal />, {
    context: {
      store,
    },
  });
  it('Should render', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
