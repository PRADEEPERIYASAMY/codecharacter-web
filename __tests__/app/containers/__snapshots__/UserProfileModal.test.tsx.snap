// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserProfileModal Container Should render 1`] = `
<UserProfileModal
  checkUsernameExists={[Function]}
  editUserPassword={[Function]}
  editUserProfile={[Function]}
  getUserDetails={[Function]}
  isUserProfileModalOpen={false}
  toggleUserProfileModal={[Function]}
  userDetails={
    Object {
      "country": "IN",
      "email": "",
      "errorMessage": "",
      "fullName": "",
      "isLoggedIn": true,
      "isUserProfileModalOpen": false,
      "username": "",
    }
  }
/>
`;
